pkg/util/utils.go

```diff
@@ -217,6 +220,103 @@ func CalculateIPFromCIDRs(IPAddresses []string) (int, error) {
 	return total, nil
 }
 
+func parseCIDRRange(cidr string) (startIP, endIP net.IP, err error) {
+	// TODO: confirm whether the error message is enough
+	_, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return nil, nil, err
+	}
+	startIP = ipnet.IP
+	endIP = make(net.IP, len(startIP))
+	copy(endIP, startIP)
+	for i := len(startIP) - 1; i >= 0; i-- {
+		endIP[i] = startIP[i] | ^ipnet.Mask[i]
+	}
+	return startIP, endIP, nil
+}
+
+func calculateOffsetIP(ip net.IP, offset int) (net.IP, error) {
+	ipInt := ipToUint32(ip)
+	ipInt += uint32(offset)
+	if int(ipInt) < 0 {
+		return nil, fmt.Errorf("resulting IP is less than 0")
+	}
+	if ipInt > 0xFFFFFFFF {
+		return nil, fmt.Errorf("resulting IP is greater than 255.255.255.255")
+	}
+	return uint32ToIP(ipInt), nil
+}
+
+func ipToUint32(ip net.IP) uint32 {
+	ip = ip.To4()
+	return binary.BigEndian.Uint32(ip)
+}
+
+func uint32ToIP(ipInt uint32) net.IP {
+	ip := make(net.IP, net.IPv4len)
+	binary.BigEndian.PutUint32(ip, ipInt)
+	return ip
+}
+
+func compareIP(ip1, ip2 net.IP) bool {
+	return ipToUint32(ip1) < ipToUint32(ip2)
+}
+
+func rangesAbstractRange(ranges [][]net.IP, except []net.IP) [][]net.IP {
+	// ranges: [[172.0.0.1 172.0.255.255] [172.2.0.1 172.2.255.255]]
+	// except: [172.0.100.1 172.0.100.255]
+	// return: [[172.0.0.1 172.0.100.0] [172.0.101.0 172.0.255.255] [172.2.0.1 172.2.255.255]]
+	//
```

Comment:
// ranges: [[172.0.0.1 172.0.255.255] [172.2.0.1 172.2.255.255]]
/ except: [172.0.100.1 172.0.100.255]
// return: [[172.0.0.1 172.0.100.0] [172.0.101.0 172.0.255.255] [172.2.0.1 172.2.255.255]]

The Line#266 - line#L268 are the middle steps or duplicated with #Line270 - #Line272.
I believe the #Line272 should be the final result, is it?